### PR TITLE
8343745: Only update Last Value Assertion Predicates in Loop Unrolling

### PR DESCRIPTION
**Pull request by chhagedorn**
(Note: This is a dependent PR on https://github.com/openjdk/jdk/pull/21944 which is fully reviewed but I'd prefer to integrate it on Monday and run again some testing over the weekend)

This is a follow-up feature to https://github.com/openjdk/jdk/pull/21944 which updated Loop Unrolling to use a new predicate visitor and enables this patch.

In Loop Unrolling, we only update the stride and not the init value of a loop. Thus, we actually only require to update the Last Value Assertion Predicates because the Init Value Assertion Predicates do not use `OpaqueLoopStride`. So, we also would not be required to kill the old Init Value Initialized Assertion Predicates. This patch implements that improvement.

To make this work, we need to query the associated `AssertionPredicateType` of an Assertion Predicate which is stored in the `If/RangeCheckNode` (I guess it's okay to have this additional node field in product). This was guarded by `NOT_PRODUCT` before. For this patch, I make this information available in product builds and use it to implement this feature.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343745](https://bugs.openjdk.org/browse/JDK-8343745): Only update Last Value Assertion Predicates in Loop Unrolling (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [e9161d16](https://git.openjdk.org/jdk/pull/21969/files/e9161d16a2654970dbe576b86f016e438531be38)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21969/head:pull/21969` \
`$ git checkout pull/21969`

Update a local copy of the PR: \
`$ git checkout pull/21969` \
`$ git pull https://git.openjdk.org/jdk.git pull/21969/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21969`

View PR using the GUI difftool: \
`$ git pr show -t 21969`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21969.diff">https://git.openjdk.org/jdk/pull/21969.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21969#issuecomment-2463937449)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/predicates.hpp [hunk 1]
```diff
@@ -1063,7 +1067,10 @@ class UpdateStrideForAssertionPredicates : public PredicateVisitor {
   // Predicates are inserted after the Template Assertion Predicate which ensures that we are not accidentally visiting
   // and killing a newly created Initialized Assertion Predicate here.
   void visit(const InitializedAssertionPredicate& initialized_assertion_predicate) override {
-    initialized_assertion_predicate.kill(_phase);
+    if (initialized_assertion_predicate.is_last_value()) {
+      // Only Last Value Initialized Assertion Predicates need to be killed and updated.
+      initialized_assertion_predicate.kill(_phase);
+    }
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.hpp:1073]: Only killing old Last Value Initialized Assertion Predicate

File: src/hotspot/share/opto/predicates.cpp [hunk 2]
```diff
@@ -879,6 +875,10 @@ IfTrueNode* CreateAssertionPredicatesVisitor::clone_template_and_replace_init_in
 
 // Clone the Template Assertion Predicate and set a new input for the OpaqueLoopStrideNode.
 void UpdateStrideForAssertionPredicates::visit(const TemplateAssertionPredicate& template_assertion_predicate) {
+  if (!template_assertion_predicate.is_last_value()) {
+    // Only Last Value Assertion Predicates have an OpaqueLoopStrideNode.
+    return;
+  }
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.cpp:881]: Skipping to update Init Value Template Assertion Predicate.

**Review by vnkozlov — APPROVED** [review]
Looks fine to me.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Pull request closed without merging**